protocol/xmrmaker/swap_state.go

```diff
@@ -136,31 +142,68 @@ func newSwapState(
 		walletScanHeight -= monero.MinSpendConfirmations
 	}
 
-	ctx, cancel := context.WithCancel(b.Ctx())
-	s := &swapState{
-		ctx:                 ctx,
-		cancel:              cancel,
-		Backend:             b,
-		sender:              sender,
-		offer:               offer,
-		offerExtra:          offerExtra,
-		offerManager:        om,
-		walletScanHeight:    walletScanHeight,
-		nextExpectedMessage: &net.SendKeysMessage{},
-		readyCh:             make(chan struct{}),
-		info:                info,
-		done:                make(chan struct{}),
+	ethHeader, err := b.EthClient().HeaderByNumber(b.Ctx(), nil)
+	if err != nil {
+		return nil, err
 	}
 
-	return s, nil
-}
+	// set up ethereum event watchers
+	logReadyCh := make(chan []ethtypes.Log)
+	logRefundedCh := make(chan []ethtypes.Log)
 
-func (s *swapState) lockState() {
-	s.stateMu.Lock()
-}
+	ctx, cancel := context.WithCancel(b.Ctx())
```

Comment:
It took me a while to track this down.  I wanted to understand how the event filterer go routines were terminating.  It looks like this context inherits from the global context, so it will be canceled if the global context is canceled, but this subcontext only lasts for the length of the swap, so when the swap ends, this context is canceled and the swap specific filtering go routines are terminated. Let me know if that is correct. If you have any ideas on how to make this more obvious when reading the code, please add whatever you think might help.

Reply:
yeah that's correct, a context is done when it cancels or one of its ancestor contexes is cancelled. not really sure other than comments, what do you think?

Reply:
added comments, also updated the two deferred functions to match more closely.